source/common/conn_pool/conn_pool_base.cc

```diff
@@ -279,11 +285,15 @@ ConnectionPool::Cancellable* ConnPoolImplBase::newStreamImpl(AttachContext& cont
                  result == ConnectionResult::FailedToCreateConnection),
             fmt::format("Failed to create expected connection: {}", *this));
   if (result == ConnectionResult::FailedToCreateConnection) {
-    // This currently only happens for HTTP/3 if secrets aren't yet loaded.
-    // Trigger connection failure.
-    pending->cancel(Envoy::ConnectionPool::CancelPolicy::CloseExcess);
-    onPoolFailure(nullptr, absl::string_view(), ConnectionPool::PoolFailureReason::Overflow,
-                  context);
+    // This currently only happens for HTTP/3 if 1) secrets aren't yet loaded,
+    // or 2) connect failed immediately.
+    if (!pending_streams_.empty()) {
```

Comment:
I'm not quite sure I understand why this is needed, but wasn't needed before. Can you elaborate?

Reply:
any chance we can just avoid this at the call site for (1) rather than here to avoid "sometimes done here, sometimes done elsewhere"?